File: akka-sample-cqrs-scala/src/main/scala/sample/cqrs/KeepAlive.scala
package sample.cqrs

import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Behaviors
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
import akka.cluster.typed.ClusterSingleton
import akka.cluster.typed.ClusterSingletonSettings
import akka.cluster.typed.SingletonActor

/**
 * General purpose event processor infrastructure. Not specific to the ShoppingCart domain.
 */
object KeepAlive {
  case object Probe

  def init(system: ActorSystem[_], eventProcessorEntityKey: EntityTypeKey[EventProcessor.Ping.type]): Unit = {
    val settings = EventProcessorSettings(system)
    ClusterSingleton(system).init(
      SingletonActor(KeepAlive(settings, eventProcessorEntityKey), s"keepAlive-${settings.id}")
        .withSettings(ClusterSingletonSettings(system).withRole("read-model")))
  }

  def apply(
      settings: EventProcessorSettings,
      eventProcessorEntityKey: EntityTypeKey[EventProcessor.Ping.type]): Behavior[Probe.type] = {
    Behaviors.setup { context =>
      Behaviors.withTimers { timers =>
        val sharding = ClusterSharding(context.system)

        timers.startTimerWithFixedDelay(Probe, Probe, settings.keepAliveInterval)

        Behaviors.receiveMessage { probe =>
          for (n <- 0 until settings.parallelism) {
            val eventProcessorId: String = s"${settings.tagPrefix}-$n"
            sharding.entityRefFor(eventProcessorEntityKey, eventProcessorId) ! EventProcessor.Ping
          }
          Behaviors.same
        }
      }
    }

  }
}

File: akka-sample-cqrs-java/src/main/java/sample/cqrs/KeepAlive.java
package sample.cqrs;

import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
import akka.cluster.typed.ClusterSingleton;
import akka.cluster.typed.ClusterSingletonSettings;
import akka.cluster.typed.SingletonActor;

public class KeepAlive extends AbstractBehavior<KeepAlive.Probe> {

  enum Probe {
    INSTANCE
  }

  public static void init(ActorSystem<?> system, EntityTypeKey<EventProcessor.Ping> eventProcessorEntityKey) {
    EventProcessorSettings settings = EventProcessorSettings.create(system);
    ClusterSingleton.get(system).init(
      SingletonActor.of(KeepAlive.create(settings, eventProcessorEntityKey), "keepAlive-" + settings.id)
        .withSettings(ClusterSingletonSettings.create(system).withRole("read-model")));
  }

  public static Behavior<Probe> create(
    EventProcessorSettings settings,
    EntityTypeKey<EventProcessor.Ping> eventProcessorEntityKey) {

    return Behaviors.setup(context ->
      Behaviors.withTimers(timers -> {
        timers.startTimerWithFixedDelay(Probe.INSTANCE, Probe.INSTANCE, settings.keepAliveInterval);
        return new KeepAlive(context, settings, eventProcessorEntityKey);
      }));
  }

  private final EventProcessorSettings settings;
  private final EntityTypeKey<EventProcessor.Ping> eventProcessorEntityKey;
  private final ClusterSharding sharding;

  private KeepAlive(ActorContext<Probe> context, EventProcessorSettings settings, EntityTypeKey<EventProcessor.Ping> eventProcessorEntityKey) {
    super(context);
    this.settings = settings;
    this.eventProcessorEntityKey = eventProcessorEntityKey;
    this.sharding = ClusterSharding.get(context.getSystem());
  }

  @Override
  public Receive<Probe> createReceive() {
    return newReceiveBuilder()
      .onMessage(Probe.class, p -> onProbe())
      .build();
  }

  private Behavior<Probe> onProbe() {
    for (int i = 0; i < settings.parallelism; i++) {
      String eventProcessorId = settings.tagPrefix + "-" + i;
      sharding.entityRefFor(eventProcessorEntityKey, eventProcessorId).tell(EventProcessor.Ping.INSTANCE);
    }
    return this;
  }


}
